File: src/services/geocode.js
/**
 * geocode.js (Nominatim structured + fallback)
 *
 * PURPOSE:
 * - Convert a structured address into coordinates (lat/lon)
 * - Fall back when house numbers aren't present in OSM data
 *
 * STRATEGY:
 * 1) Try full structured address (street includes house number)
 * 2) If that fails, remove leading house number and try again
 * 3) If that fails, fall back to city/state/zip area lookup
 *
 * IMPORTANT (Project 1 Dev Setup):
 * - call "/nominatim/search" instead of the full domain.
 * - Vite proxy forwards this to https://nominatim.openstreetmap.org
 * - This avoids CORS issues in the browser.
 */

const NOMINATIM_BASE = "/nominatim/search";

async function nominatimSearch(paramsObj) {
  const params = new URLSearchParams({
    format: "json",
    limit: "1",
    addressdetails: "1",
    ...paramsObj,
  });

  const url = `${NOMINATIM_BASE}?${params.toString()}`;

  const res = await fetch(url, {
    headers: {
      Accept: "application/json",
      // NOTE:
      // We do NOT set "User-Agent" here because browsers forbid it.
      // If using a backend proxy in production, User-Agent should be set there.
    },
  });

  if (!res.ok) {
    throw new Error(`Geocoding failed (${res.status})`);
  }

  return res.json();
}

function pickFirst(data) {
  if (!Array.isArray(data) || data.length === 0) return null;
  const first = data[0];
  return {
    lat: Number(first.lat),
    lon: Number(first.lon),
    label: first.display_name || "",
  };
}

function removeLeadingHouseNumber(street) {
  return String(street || "").trim().replace(/^\s*\d+\s+/, "");
}

/**
 * geocodeStructured
 * - addressParts: { street, city, state, zip }
 * - returns: { lat, lon, label, usedFallback }
 */
export async function geocodeStructured(addressParts) {
  const street = String(addressParts.street || "").trim();
  const city = String(addressParts.city || "").trim();
  const state = String(addressParts.state || "").trim().toUpperCase();
  const zip = String(addressParts.zip || "").trim();

  // 1) Full structured (best accuracy)
  let data = await nominatimSearch({
    street,
    city,
    state,
    postalcode: zip,
    countrycodes: "us",
  });

  let hit = pickFirst(data);
  if (hit) return { ...hit, usedFallback: false };

  // 2) Fallback: remove house number from street (street-level geocode)
  const streetNoNumber = removeLeadingHouseNumber(street);
  if (streetNoNumber && streetNoNumber !== street) {
    data = await nominatimSearch({
      street: streetNoNumber,
      city,
      state,
      postalcode: zip,
      countrycodes: "us",
    });

    hit = pickFirst(data);
    if (hit) return { ...hit, usedFallback: true };
  }

  // 3) Fallback: area centroid using city/state/zip (route still possible)
  data = await nominatimSearch({
    city,
    state,
    postalcode: zip,
    countrycodes: "us",
  });

  hit = pickFirst(data);
  if (hit) return { ...hit, usedFallback: true };

  throw new Error(
    `No geocoding results for: "${street}, ${city}, ${state} ${zip}". Try checking spelling or using a nearby landmark.`
  );
}

/**
 * geocodeAddress
 * - Free-form fallback (string)
 * - returns: { lat, lon, label, usedFallback }
 */
export async function geocodeAddress(address) {
  const q = String(address || "").trim();
  const data = await nominatimSearch({ q, countrycodes: "us" });
  const hit = pickFirst(data);

  if (!hit) {
    throw new Error(`No geocoding results for: "${q}"`);
  }

  // Free-form geocode is considered fallback behavior
  return { ...hit, usedFallback: true };
}

File: server/app/services/geocode.py
import requests

NOMINATIM_BASE = "https://nominatim.openstreetmap.org/search"

HEADERS = {
    "Accept": "application/json",
    "User-Agent": "OnTimePlanner/0.1 (student project)",
    "Accept-Language": "en",
}

def _search(params):
    res = requests.get(NOMINATIM_BASE, params=params, headers=HEADERS, timeout=15)
    res.raise_for_status()
    return res.json()

def _pick_first(data):
    if not isinstance(data, list) or len(data) == 0:
        return None
    first = data[0]
    return {
        "lat": float(first["lat"]),
        "lon": float(first["lon"]),
        "label": first.get("display_name", ""),
    }

def _remove_leading_house_number(street):
    import re
    return re.sub(r"^\s*\d+\s+", "", str(street or "").strip())

def geocode_structured(parts):
    street = str(parts.get("street", "")).strip()
    city = str(parts.get("city", "")).strip()
    state = str(parts.get("state", "")).strip().upper()
    zip_code = str(parts.get("zip", "")).strip()

    data = _search({
        "format": "json",
        "limit": 1,
        "addressdetails": 1,
        "street": street,
        "city": city,
        "state": state,
        "postalcode": zip_code,
        "countrycodes": "us",
    })

    hit = _pick_first(data)
    if hit:
        return {**hit, "usedFallback": False}

    street_no_number = _remove_leading_house_number(street)
    if street_no_number and street_no_number != street:
        data = _search({
            "format": "json",
            "limit": 1,
            "addressdetails": 1,
            "street": street_no_number,
            "city": city,
            "state": state,
            "postalcode": zip_code,
            "countrycodes": "us",
        })
        hit = _pick_first(data)
        if hit:
            return {**hit, "usedFallback": True}

    data = _search({
        "format": "json",
        "limit": 1,
        "addressdetails": 1,
        "city": city,
        "state": state,
        "postalcode": zip_code,
        "countrycodes": "us",
    })
    hit = _pick_first(data)
    if hit:
        return {**hit, "usedFallback": True}

    raise ValueError(f'No geocoding results for "{street}, {city}, {state} {zip_code}"')
